src/download_px4_logfile/main.go

```diff
@@ -0,0 +1,285 @@
+}
+
+func (s *LogItemStore) Add(dir string, file mavctrl.Diritem) {
+	base, ext := SplitFileName(file.Name)
+	//log.Println("[LogItemStore] Add: " + base + " " + ext)
+	switch ext {
+	case "ulg":
+		// Non-encrypted log file
+		var item LogFileItem
+		item.Id = len(s.Items)
+		item.Dir = dir
+		item.Crypted = false
+		item.DataFile = file
+		s.Items = append(s.Items, item)
+	case "ulgc":
+		// Enqrypted log data file
+		item_p, found := s.ItemExists(dir, base)
+		if found {
+			item_p.DataFile = file
+		} else {
+			var item LogFileItem
+			item.Id = len(s.Items)
+			item.Dir = dir
+			item.Crypted = true
+			item.DataFile = file
+			s.Items = append(s.Items, item)
+		}
+	case "ulgk":
+		// Enqrypted log key file
+		item_p, found := s.ItemExists(dir, base)
+		if found {
+			item_p.KeyFile = file
+		} else {
+			var item LogFileItem
+			item.Id = len(s.Items)
+			item.Dir = dir
+			item.Crypted = true
+			item.KeyFile = file
+			s.Items = append(s.Items, item)
+		}
+	}
+}
+
+func (s *LogItemStore) GetLatestId() (int, bool) {
+	var latest_time time.Time
+	var latest_id int = -1
+	found := false
+
+	for i := 0; i < len(s.Items); i++ {
+		base, _ := SplitFileName(s.Items[i].DataFile.Name)
+		tm, err := time.Parse("2006-01-02 15_04_05", s.Items[i].Dir+" "+base)
+		if err == nil {
+			if latest_id < 0 || latest_time.Before(tm) {
+				latest_time = tm
+				latest_id = i
+				found = true
+			}
+		}
+	}
+
+	return latest_id, found
+}
+
+func (s *LogItemStore) FetchLog(sel int) {
+	if sel >= 0 && sel < len(s.Items) {
+		dir := s.Items[sel].Dir
+		name := s.Items[sel].DataFile.Name
+		size := s.Items[sel].DataFile.Size
+		base, ext := SplitFileName(name)
+		outpath_no_ext := *argsFlightLogDirPtr + "/log-" + dir + "T" + strings.Replace(base, "_", "-", 0) + "Z."
+		log.Printf("[FetchLog] : %s/%s (%d bytes)\n", dir, name, size)
+		ctrl.DownloadLogFile(dir, name, outpath_no_ext+ext, size)
+		if s.Items[sel].Crypted {
+			size = s.Items[sel].KeyFile.Size
+			name := s.Items[sel].KeyFile.Name
+			_, ext = SplitFileName(name)
+			log.Printf("[FetchLog] : %s/%s (%d bytes)\n", dir, name, size)
+			ctrl.DownloadLogFile(dir, name, outpath_no_ext+ext, size)
+		}
+	}
+}
+
+/*********************************************
+ * Main
+ */
+
+func logic() bool {
+	var success bool = true
+	flag.Parse()
+	if *argsVersionPtr {
+		fmt.Println("version:", getVersion())
+		return success
+	}
+
+	log.Println("Flight-log-downloader, version:", getVersion())
+	if *argsFlightLogDirPtr == "." {
+		path, err := os.Getwd()
+		if err == nil {
+			*argsFlightLogDirPtr = path
+		}
+	}
+	log.Println("Flight log dir: ", *argsFlightLogDirPtr)
+
+	// Initialize
+	if !ctrl.Init(*argsAddressPtr) {
+		log.Fatalf("Can't connect to Flight Controller via Mavlink ('%s')\n", *argsAddressPtr)
+	}
+	defer ctrl.Close()
+
+	if !ctrl.CheckLogDirExists() {
+		return false
+	}
+
+	var log_items LogItemStore
+
+	log_dir := "/fs/microsd/log"
+	log_dir_list, ok := ctrl.ListDirectory("/fs/microsd/log")
+	if ok {
```

Comment:
Regarding my other comment: https://github.com/tiiuae/fogsw_tools/pull/22/files#r949856082

This looks also like a critical error, i.e. if directory listing fails, I would expect it to be a critical error in context of downloading log files.

So I think it would make sense to stop program execution and handle it the same as in the referenced comment.

Also the negation (`if err != nil { return err }`) would reduce one level of indentation.